cmd/pfaFindFast/efficient.go

```diff
@@ -0,0 +1,337 @@
+						exception.PanicOnErr(err)
+					}
+				}
+			}
+		}
+	}
+
+	// close outFile
+	err = file.Close()
+	exception.PanicOnErr(err)
+}
+
+// distWindow takes 2 pDna sequences (aligned in a multi-pFa format), and window start and end positions as inputs.
+// The function calculates the Euclidean distance between the 2 pDna base vectors at the same position in the 2 pDna sequences for each position in the window.
+// The function returns both the slice containing all the distance calculations within the window, and the mean.
+func distWindow(firstQuery []pDna.Float32Base, secondQuery []pDna.Float32Base, windowStart int, windowEnd int) (windowDist []float64, windowDistMean float64) {
+	var baseDist float64
+	var windowDistTotal = 0.0
+
+	for i := windowStart; i <= windowEnd; i++ {
+		baseDist = pDna.Dist(firstQuery[i], secondQuery[i])
+		windowDist = append(windowDist, baseDist) // appending each baseDist instead of defining windowDist size first
+		windowDistTotal += baseDist
+	}
+
+	windowDistMean = windowDistTotal / float64(windowEnd-windowStart+1) // divide by actual windowSize in reference, not input windowSize in query sequences
+
+	return windowDist, windowDistMean
+}
+
+// windowDistToDiv takes 2 pDna sequences (aligned in a multi-pFa format), window start and end positions, and a user-specified float64 threshold as inputs.
+// The function counts the number of divergences between the 2 pDna sequences within the window.
+// The rules for counting divergences are as follows:
+// Gaps are not counted as divergences.
+// If the Euclidean distance between the 2 pDna base vectors at the same position in the 2 pDna sequences is greater than the threshold, then a divergence is counted.
+func windowDistToDiv(firstQuery []pDna.Float32Base, secondQuery []pDna.Float32Base, windowStart int, windowEnd int, baseDistToDivThreshold float64) int {
+	var baseDist float64
+	var windowDistDiv = 0
+
+	for i := windowStart; i <= windowEnd; i++ {
+		if pDna.IsGap(firstQuery[i]) || pDna.IsGap(secondQuery[i]) { // do not count gaps
+			continue
+		} else {
+			baseDist = pDna.Dist(firstQuery[i], secondQuery[i])
+			if baseDist > baseDistToDivThreshold {
+				windowDistDiv++ // baseDistDiv is true, add 1 to the windowDistDiv count
+			}
+		}
+	}
+
+	return windowDistDiv
+}
+
+// dotWindow takes 2 pDna sequences (aligned in a multi-pFa format), and window start and end positions as inputs.
+// The function calculates the dot product distance (1 - dot product) between the 2 pDna base vectors at the same position in the 2 pDna sequences for each position in the window.
+// The function returns both the slice containing all the distance calculations within the window, and the mean.
+func dotWindow(firstQuery []pDna.Float32Base, secondQuery []pDna.Float32Base, windowStart int, windowEnd int) (windowDot []float64, windowDotMean float64) {
```

Comment:
Remove dotWindow and distWindow since they don't contribute to substitution/divergence count